File: frontend/src/lib/nostr/roundrobin.ts
// Pure fair round-robin scheduling (no Nostr/state) — a mirror of relay/roundrobin.go (fairNext),
// used only for the client's read-only "Up next" preview. The relay is the authority.
//
// DJs take turns starting AFTER the currently-playing DJ; each contributes its TOP playable track
// (first ACTIVE, not the currently-playing one). So the interleave alternates fairly by the k-th
// PLAYABLE track of each DJ — never by absolute queue index — regardless of where each DJ's `off`
// tracks sit. The visible queue is the single source of truth: a played track is `off` and drops
// out, a reorder changes the top, a re-activated track returns. No hidden played-set.

export interface Slot {
  djIndex: number
  trackIndex: number
}

/**
 * The next `count` tracks the round-robin will play, as (djIndex, trackIndex) slots.
 * `playable[i][j]` = true when DJ i's track j is ACTIVE (not `off`) and not the currently-playing
 * one. `lastDjIndex` = index of the currently-playing DJ (or -1 to start from dj0). Looks ahead by
 * simulating consumption (a picked track won't be picked again), so each DJ contributes its next
 * active track each round — exactly the order the relay's repeated `advance` produces. Stops early
 * when nothing is left.
 */
export function fairSequence(
  djCount: number,
  playable: boolean[][],
  lastDjIndex: number,
  count: number,
): Slot[] {
  const out: Slot[] = []
  if (djCount === 0) return out
  const consumed = playable.map((row) => row.map(() => false))
  let last = lastDjIndex
  for (let k = 0; k < count; k++) {
    let picked = -1
    let pickedTi = -1
    const start = last + 1 // -1 (none playing) → start at dj0
    for (let o = 0; o < djCount; o++) {
      const di = (((start + o) % djCount) + djCount) % djCount
      const row = playable[di] ?? []
      for (let ti = 0; ti < row.length; ti++) {
        if (row[ti] && !consumed[di][ti]) {
          picked = di
          pickedTi = ti
          break
        }
      }
      if (picked >= 0) break
    }
    if (picked < 0) break
    consumed[picked][pickedTi] = true
    out.push({ djIndex: picked, trackIndex: pickedTi })
    last = picked
  }
  return out
}

File: relay/roundrobin.go
package main

// Fair round-robin scheduling — picks the next track to play across the stage DJs.
//
// Each advance, the DJs take turns (round-robin) starting right AFTER the DJ who just played.
// Every DJ contributes its TOP playable track — the first ACTIVE (not `off`) track in its queue
// that isn't the currently-playing one. So the interleave alternates fairly across DJs by the
// k-th PLAYABLE track of each — never by absolute queue index — regardless of where each DJ's
// `off` (played/disabled) tracks happen to sit. The visible queue is the single source of truth:
// a played track becomes `off` and drops out (the DJ's "top" moves down), a reorder changes the
// top immediately, a re-activated track returns to the rotation. No hidden played-set. Mirrors
// frontend/src/lib/nostr/roundrobin.ts (fairSequence) — the relay is the authority, the TS copy
// drives the client's "Up next" preview only.

// fairNext picks the next (djIndex, trackIndex) to play: round-robin across djPks starting after
// lastDJ, each DJ offering its first playable track. playable[i][j] = true when DJ i's track j is
// ACTIVE and not the currently-playing one. Returns (-1, -1) when nothing is playable.
func fairNext(djPks []string, lastDJ string, playable [][]bool) (djIndex, trackIndex int) {
	n := len(djPks)
	if n == 0 {
		return -1, -1
	}
	start := indexOf(djPks, lastDJ) + 1 // -1 (not found / empty lastDJ) → start at dj0
	for off := 0; off < n; off++ {
		di := (start + off) % n
		if di >= len(playable) {
			continue
		}
		for ti, ok := range playable[di] {
			if ok {
				return di, ti
			}
		}
	}
	return -1, -1
}

func indexOf(s []string, v string) int {
	for i, x := range s {
		if x == v {
			return i
		}
	}
	return -1
}
